### PR TITLE
Windows: FlutterDesktopPixelBuffer: Add optional release callback

### PR DESCRIPTION
**Pull request by jnschulze**
This PR adds an optional callback to `FlutterDesktopPixelBuffer` that gets invoked when it's safe to release the backing buffer (or unlock a mutex protecting that buffer etc.).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by jnschulze** [comment]
@chinmaygarde

**Review by chinmaygarde — APPROVED** [review]

**Comment by chinmaygarde** [comment]
I think this is a fine addition. But can we a test for this? I don't see an existing test for this so I am unable to provide any guidance here. Perhaps @cbracken might know.

**Comment by cbracken** [comment]
Will take a quick look -- I'm pretty sure I added a test for something similar a couple months back.

**Comment by chinmaygarde** [comment]
Chris may be busy with other things. To test this, I think we can just add a latch that is signalled in the unregistration callback and wait for it after calling `FlutterEngineUnregisterExternalTexture`. If the test times out, the watchdog should terminate the harness anyway. Since the test seems easy-ish to wire up, let's attempt to write one before landing this.

**Comment by jnschulze** [comment]
@chinmaygarde 
I've mocked all gl* functions used by `ExternalTextureGL` and added a test for `ExternalTextureGL::PopulateTexture` (which invokes the newly introduced callback after the pixel buffer has been copied).
So we not only test the new callback but also the texture population itself.

**Review by chinmaygarde — APPROVED** [review]

**Comment by chinmaygarde** [comment]
Thanks!

**Review by cbracken — COMMENTED** [review]
lgtm! Thanks!

File: shell/platform/common/public/flutter_texture_registrar.h [hunk 1]
```diff
@@ -34,6 +34,10 @@ typedef struct {
   size_t width;
   // Height of the pixel buffer.
   size_t height;
+  // An optional callback that gets invoked when the |buffer| can be released.
+  void (*release_callback)(void* release_context);
+  // Opaque data passed to |release_callback|.
+  void* release_context;
```
Comment (stuartmorgan) [inline shell/platform/common/public/flutter_texture_registrar.h:40]: Adding these caused undefined behavior for existing code (unless the code was specifically written to defend against the possibility of new values by zero-initializing the entire memory block).

If we edit this again we should add a struct size to do safe reads, as we do with embedder.h.

**Comment by MindStudioOfficial** [comment]
I have set up the Following buffer update function:

```cpp
void Frame::Update(uint8_t *buffer, int32_t width, int32_t height)
{
    const std::lock_guard<std::mutex> lock(mutex_);
    flutter_pixel_buffer_.buffer = buffer;
    flutter_pixel_buffer_.width = width;
    flutter_pixel_buffer_.height = height;
    flutter_pixel_buffer_.release_context = &buffer;
    flutter_pixel_buffer_.release_callback = [](void *user_data)
    {
        uint8_t *old_buffer = reinterpret_cast<uint8_t *>(user_data);
        if (old_buffer != nullptr)
            free(old_buffer);
    };
    texture_registrar_->MarkTextureFrameAvailable(texture_id_);
}
```

This gets invoked by dart throught the method channel

However when I try to `free(old_buffer)`

I get the following error:

```
Debug Assertion Failed!
...
Expression: _CrtlsValidHeapPointer(block)
...
```

The original buffer is a `uint8_t *` from dart ffi.

How and where am I supposed to free the buffer?

**Pull request merged**